### PR TITLE
Change AttributeBasedFormatter to allow custom formatter selection ba…

### LINKED ISSUE #309: Allow custom formatters to "kick in" for more specific types
Custom formatters (I am referring to Formatters section in https://github.com/dennisdoomen/fluentassertions/wiki#extensibility) are right now selected by exact match of types:

https://github.com/dennisdoomen/fluentassertions/blob/develop/Src/Core/Formatting/AttributeBasedFormatter.cs#L63

Given #308 I wonder if it would be better to go by `isAssignableFrom()` instead? It would then be possible to implement a _catch all_ custom formatter with `object` as an argument and completely disable the default formatter that builds the object's graph.

Or maybe introduce an interface that formatters would need to implement to give them (the formatters) the ability to respond to `CanHandle()` ?

### PR DESCRIPTION
**Pull request by Serg046**
…sed on the parent type

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution adds a feature or fixes a bug, please update [**the release notes**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which are published on the [website](https://fluentassertions.com/releases).
* [x] If the contribution changes the public API the changes needs to be included by running `AcceptApiChanges.ps1`/`AcceptApiChanges.sh`.
* [ ] If the contribution affects [the documentation](https://github.com/fluentassertions/fluentassertions/tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

Fixes #309

### CODE REVIEW

File: Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs [hunk 1]
```diff
@@ -46,9 +46,34 @@ public string Format(object value, FormattingContext context, FormatChild format
...
             Type valueType = value.GetType();
-            MethodInfo formatter = Formatters.FirstOrDefault(m => m.GetParameters().Single().ParameterType == valueType);
+            var possibleFormatters = Formatters
+                .Select(m => new
+                {
+                    Type = m.GetParameters().Single().ParameterType,
+                    Method = m
+                })
+                .Where(f => valueType.IsSameOrInherits(f.Type))
+                .ToDictionary(f => f.Type, f => f.Method);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:56]: 🤔 Is there a need to order the formatters in some way to get some deterministic order?

Comment (Serg046) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:56]: That would require multiple enumerations and I guess wouldn't improve readability.
One enumeration using do/while looks more efficient and readable to me.

Comment (dennisdoomen) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:58]: Imagine multiple classes implementing the same interface `IFoo`. What would happen if you two formatters, one that accepts a `IFoo` and one, more specialized, that takes an implementation of `IFoo`.

Comment (jnyrup) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:58]: @Serg046 if efficiency is a concern here, could you then cache the `Dictionary` just as `formatters` is?

Comment (Serg046) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs]: Did it but forced to change one existing test a bit because of cache.

**Review by dennisdoomen — APPROVED** [review]

File: Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs [hunk 2]
```diff
@@ -46,11 +47,45 @@ public string Format(object value, FormattingContext context, FormatChild format
...
+            {
+                var possibleFormatters = Formatters
+                    .Select(m => new { Type = m.GetParameters().Single().ParameterType, Method = m })
+                    .GroupBy(f => f.Type)
+                    .Select(g => g.First())
+                    .Where(f => valueType.IsSameOrInherits(f.Type))
+                    .ToDictionary(f => f.Type, f => f.Method);
+
+                formatter = GetFormatter(possibleFormatters, valueType);
+                determinedFormatters.Add(valueType, formatter);
```
Comment (jnyrup) [inline Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:60]: Sorry for being too vague on my thoughts about caching.

As `AttributeBasedFormatter` is instantiated a single time in the static `Formatter.DefaultFormatters` we cannot use `Dictionary<TKey, TValue>` as a cache as `Add` is not thread-safe.

My idea is to change 
```cs
private MethodInfo[] formatters;
``` 
into
```cs
private Dictionary<Type, MethodInfo> formatters;
```

```cs
private Dictionary<Type, MethodInfo> Formatters
{
    get
    {
        HandleValueFormatterDetectionModeChanges();

        return formatters ??= GetDictionary(FindCustomFormatters());
    }
}

// with a better name or merge into FindCustomFormatters
private static Dictionary<Type, MethodInfo> GetDictionary(MethodInfo[] formatters)
{
    return formatters
        .Select(m => new { Type = m.GetParameters().Single().ParameterType, Method = m })
        .GroupBy(f => f.Type)
        .Select(g => g.First())
        .ToDictionary(f => f.Type, f => f.Method);
}
```

```cs
private MethodInfo GetFormatter(object value)
{
    Type valueType = value.GetType();
    return GetFormatter(valueType);
}

private MethodInfo GetFormatter(Type valueType)
{
    do
    {
        if (Formatters.TryGetValue(valueType, out var formatter))
        {
            return formatter;
        }

        valueType = valueType.BaseType;
    }
    while (valueType != null);

    return null;
}
```

Comment (Serg046) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:60]: Thanks @jnyrup, got it.
You miss `IsSameOrInherits` check in your example. How do you suggest including it?

Comment (jnyrup) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs:60]: I deliberately ignored `IsSameOrInherits` 😀 
I might be missing something, but I couldn't see it was needed?

Comment (Serg046) [reply Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs]: There was a glitch in my head, I was confused by `possibleFormatters.Count == 1` when we didn't need to analyze type. But I don't think this is really needed. Finally pushed your example.
Just wondering why AttributeBasedFormatter is public.

File: Tests/FluentAssertions.Specs/Formatting/FormatterSpecs.cs [hunk 3]
```diff
@@ -586,22 +586,79 @@ public void When_a_custom_formatter_exists_in_any_loaded_assembly_it_should_over
...
+
+            // Act
+            string result = Formatter.ToString(subject);
+
+            // Assert
+            result.Should().Be("Property is SomeValue", "it should use my custom formatter");
+        }
+
         [Fact]
         public void When_no_custom_formatter_exists_in_the_specified_assembly_it_should_use_the_default()
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Formatting/FormatterSpecs.cs:644]: I'm not sure why this method is changed?

Comment (Serg046) [reply Tests/FluentAssertions.Specs/Formatting/FormatterSpecs.cs:644]: Reverted it back, thank you...
It was needed because of old cache.

Comment (jnyrup) [reply Tests/FluentAssertions.Specs/Formatting/FormatterSpecs.cs:644]: Thanks for the clarification 👍

**Review by jnyrup — APPROVED** [review]
@dennisdoomen Are you fine with the API breakage?

**Comment by dennisdoomen** [comment]
Which breakage are you talking about?

**Comment by Serg046** [comment]
AttributeBasedFormatter is public.
Since we are changing `Formatter` from array to dictionary anyway, I changed the member to private.

**Comment by jnyrup** [comment]
@dennisdoomen 
> Which breakage are you talking about?

Removing `Formatters` from the public type `AttributeBasedFormatter`.
The property has been public since the class was introduced, but it seems more of an implementation detail.

**Comment by dennisdoomen** [comment]
I'm fine with that.

**Comment by jnyrup** [comment]
@Serg046 As a last thing, can I ask you to add a line to `releases.md`?

**Comment by jnyrup** [comment]
@Serg046 Thanks for another nice  contribution and for going through the iterations it took to get to this result.

**Pull request merged**